xdsl/dialects/riscv.py

```diff
@@ -232,6 +232,8 @@ def a_register(cls, index: int) -> Self:
 
 RDInvT = TypeVar("RDInvT", bound=RISCVRegisterType)
 RSInvT = TypeVar("RSInvT", bound=RISCVRegisterType)
+RS1InvT = TypeVar("RS1InvT", bound=RISCVRegisterType)
+RS2InvT = TypeVar("RS2InvT", bound=RISCVRegisterType)
```

Comment:
If we want the ability for RS1 and RS2 to be different then they must be different type variables, since substitution happens at the type variable directly, as opposed to the generic parameter (like in every other generic type system I'm aware of).